File: 894.all-possible-full-binary-trees.go
// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfs(n int, memo map[int][]*TreeNode) []*TreeNode {
	if v, ok := memo[n]; ok {
		return v
	}
	res := []*TreeNode{}
	if n%2 == 0 {
		memo[n] = res
		return res
	}
	if n == 1 {
		memo[1] = []*TreeNode{{Val: 0}}
		return memo[1]
	}
	n = n - 1
	if v, ok := memo[n]; ok {
		return v
	}
	for i := 1; i < n; i += 2 {
		left := dfs(i, memo)
		right := dfs(n-i, memo)
		for _, l := range left {
			for _, r := range right {
				res = append(res, &TreeNode{Val: 0, Left: l, Right: r})
			}
		}
	}
	memo[n] = res
	return res
}

func allPossibleFBT(n int) []*TreeNode {
	return dfs(n, map[int][]*TreeNode{})
}

// @leet end

File: 894.all-possible-full-binary-trees.kt
// @leet start
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun dfs(n: Int, memo: MutableMap<Int, List<TreeNode?>>): List<TreeNode?> {
      if (memo.containsKey(n)) { return memo[n]!! }
      val res = mutableListOf<TreeNode?>()
      if (n % 2 == 0) {
        memo[n] = res
        return res
      }
      if (n == 1) {
        res.add(TreeNode(0))
      }
      val m = n - 1
      if (memo.containsKey(m)) { return memo[m]!! }
      for (i in 1..m step 2) {
        val left = dfs(i, memo)
        val right = dfs(m - i, memo)
        for (l in left) {
          for (r in right) {
            res.add(TreeNode(0, l, r))
          }
        }
      }
      memo[m] = res
      return res
    }
    fun allPossibleFBT(n: Int): List<TreeNode?> {
      return dfs(n, mutableMapOf())
    }
}
// @leet end
